### DISCUSSION TITLE
Middleware to always use 'shallow' or other EqualityChecker if not defined otherway?

### DISCUSSION THREAD

**Discussion by laurib** [pmndrs/zustand#381] [Q&A]
I usually always need 'shallow' for comparison, but i can't figure out how to make correctly typed middleware.

I have been using enhancer function with zustand 2:

```
const enhancer = (useStore) => (selector,equalityFn) => {
  if (selector) {
    return useStore(selector, typeof equalityFn !== 'undefined' ? equalityFn : shallow);
   }
   return useStore();
};
```

But now updating to zustand 3, i cant figure out how to type it and is it correct way to do it? 
Maybe it should be done with middleware ?

Can someone help ?

**Comment by dai-shi** [comment]
Instead of struggling with store creator, would you make a custom hook?
```js
const useStoreOrig = create(...)
const useStore = (selector, equalityFn = shallow) => useStoreOrig(selector, equalityFn)
```
I think it would be easier to type.

**Reply by laurib** [reply]
How would you type this useStoreShallow ?

```
const createMyStore = <S extends State, U extends ActionCreators<S>>({
  defaultState,
  reducer,
  actions,
}: StoreOptions<S, U>): Store<S, U> => {
  const useStore = create<S>(immer(redux(reducer, defaultState)));
  const useStoreShallow = (selector, equalityFn = shallow) => useStore(selector, equalityFn);
  const boundActions = bindActionsCreators<S, U>(actions, useStore.dispatch);
  const store: Store<S, U> = [useStoreShallow, boundActions];
  return store;
};
```

**Reply by dai-shi** [reply]
Just a guess, but would this work?
```ts
  const useStoreShallow = <T>(selector: StateSelector<S, T>, equalityFn: EqualityChecker<T> = shallow) => useStore(selector, equalityFn);
```

**Reply by laurib** [reply]
i get this kind of error: 

`Type '<T>(selector: StateSelector<S, T>, equalityFn?: EqualityChecker<T>) => T' is missing the following properties from type 'UseStore<S>': setState, getState, subscribe, destroy, dispatch`

**Reply by dai-shi** [reply]
Do you want to export all those apis? Then, we'd need something like `Object.assign(useStoreShallow, useStore)` and typing becomes a bit tricky.

Otherwise, you want to change the type of `Store<S, U>`. Not sure how it's defined.

**Reply by laurib** [reply]
Yes. I'd like to use all those api calls. 
Well i possibly could do something like  const [useMyStore, actions, api ] = createStore(....) where useMyStore is `useStoreShallow` and api is `useStore` ? But it could be nice using useStoreShallow for api calls


My Store type
`
export type Store<S, U extends ActionCreators<S>> = [UseStore<S>, Actions<U>];`

**Reply by dai-shi** [reply]
```js
const [useMyStore, actions, api ] = createStore(....)
```
Well, I personally like this.

Anyway, if you want the same signature, you need Object.assign and in this case temporary `any` would be just easier.

```ts
const useStoreShallow: any = (selector, equalityFn) => useStore(selector, equalityFn || selector && shallow);
Object.assign(useStoreShallow, useStore)
```
I tweaked to support non-selector version. (Now, this gets complicated more than I first expected.)
